p2p/gossip/priorityq.go

```diff
@@ -0,0 +1,93 @@
+package gossip
+
+import (
+	"errors"
+)
+
+var (
+	ErrAlreadySet  = errors.New("a queue is already associated with this name")
+	ErrNotExist    = errors.New("no queue associated with this name")
+	ErrQueueClosed = errors.New("the queue is closed")
+)
+
+type PriorityQ struct {
+	waitCh chan struct{}
+	prios  map[string]int
+	queues []chan interface{}
+}
+
+func NewPriorityQ(queueCount int) *PriorityQ {
+	return &PriorityQ{
+		waitCh: make(chan struct{}),
+		prios:  make(map[string]int, queueCount),
+		queues: make([]chan interface{}, queueCount),
+	}
+}
+
+// Set the name of the queue with to the provided priority and buffer size
+// Returns an error if the name is already associated with a queue
+// Note: all calls to Set are assumed to be called before any call to Read/Write
+// Note: multiple names may be associated with the same priority
+func (pq *PriorityQ) Set(name string, prio int, bufferSize int) error {
```

Comment:
this impl seems odd to me, why not just pass prio to write ?

Reply:
Following our conversation:
1. PrioQ's Set method should expose priority (from a predefined set of priorities)
2. To avoid coupling, prio will be passed to the write method. This way the mapping of protocol->prio can be kept in protocol.go.